File: specs/typespec/agency/agency-users.go
package agency

import (
	"errors"
	"strings"

	"vetchium-api-server.typespec/common"
)

// Token types
type AgencySessionToken string
type AgencyTFAToken string
type AgencyInvitationToken string
type DNSVerificationToken string
type AgencySignupToken string

// RBAC types (re-exported from common)
type RoleName = common.RoleName
type AssignRoleRequest = common.AssignRoleRequest
type RemoveRoleRequest = common.RemoveRoleRequest

// ============================================
// Signup Flow (DNS-based Domain Verification)
// ============================================

type AgencyInitSignupRequest struct {
	Email      common.EmailAddress `json:"email"`
	HomeRegion string              `json:"home_region"`
}

func (r AgencyInitSignupRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError

	if r.Email == "" {
		errs = append(errs, common.NewValidationError("email", common.ErrRequired))
	} else if err := common.ValidateEmployerEmail(r.Email); err != nil {
		// Use employer email validation which blocks personal email domains
		errs = append(errs, common.NewValidationError("email", err))
	}

	if r.HomeRegion == "" {
		errs = append(errs, common.NewValidationError("home_region", common.ErrRequired))
	}

	return errs
}

type AgencyInitSignupResponse struct {
	Domain         common.DomainName `json:"domain"`
	DNSRecordName  string            `json:"dns_record_name"`
	TokenExpiresAt string            `json:"token_expires_at"`
	Message        string            `json:"message"`
}

type AgencyGetSignupDetailsRequest struct {
	SignupToken AgencySignupToken `json:"signup_token"`
}

func (r AgencyGetSignupDetailsRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError

	if r.SignupToken == "" {
		errs = append(errs, common.NewValidationError("signup_token", common.ErrRequired))
	}

	return errs
}

type AgencyGetSignupDetailsResponse struct {
	Domain common.DomainName `json:"domain"`
}

// AgencyCompleteSignupRequest completes agency signup after DNS verification.
// The first user is automatically granted admin rights and assigned
// the 'agency:superadmin' role.
// All operations are atomic - either the entire signup succeeds or no data is created.
type AgencyCompleteSignupRequest struct {
	SignupToken       AgencySignupToken   `json:"signup_token"`
	Password          common.Password     `json:"password"`
	PreferredLanguage common.LanguageCode `json:"preferred_language"`
	HasAddedDNSRecord bool                `json:"has_added_dns_record"`
	AgreesToEULA      bool                `json:"agrees_to_eula"`
}

var (
	errDNSRecordNotConfirmed = errors.New("You must confirm that you have added the DNS record")
	errEULANotAccepted       = errors.New("You must agree to the End User License Agreement")
)

func (r AgencyCompleteSignupRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError

	if r.SignupToken == "" {
		errs = append(errs, common.NewValidationError("signup_token", common.ErrRequired))
	}

	if r.Password == "" {
		errs = append(errs, common.NewValidationError("password", common.ErrRequired))
	} else if err := r.Password.Validate(); err != nil {
		errs = append(errs, common.NewValidationError("password", err))
	}

	if r.PreferredLanguage == "" {
		errs = append(errs, common.NewValidationError("preferred_language", common.ErrRequired))
	}

	if !r.HasAddedDNSRecord {
		errs = append(errs, common.NewValidationError("has_added_dns_record", errDNSRecordNotConfirmed))
	}

	if !r.AgreesToEULA {
		errs = append(errs, common.NewValidationError("agrees_to_eula", errEULANotAccepted))
	}

	return errs
}

type AgencyCompleteSignupResponse struct {
	SessionToken AgencySessionToken `json:"session_token"`
	AgencyUserID string             `json:"agency_user_id"`
}

// ============================================
// Login Flow
// ============================================

type AgencyLoginRequest struct {
	Email    common.EmailAddress `json:"email"`
	Domain   common.DomainName   `json:"domain"`
	Password common.Password     `json:"password"`
}

func (r AgencyLoginRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError

	if r.Email == "" {
		errs = append(errs, common.NewValidationError("email", common.ErrRequired))
	} else if err := r.Email.Validate(); err != nil {
		errs = append(errs, common.NewValidationError("email", err))
	}

	if r.Domain == "" {
		errs = append(errs, common.NewValidationError("domain", common.ErrRequired))
	} else if err := r.Domain.Validate(); err != nil {
		errs = append(errs, common.NewValidationError("domain", err))
	}

	if r.Password == "" {
		errs = append(errs, common.NewValidationError("password", common.ErrRequired))
	} else if err := r.Password.Validate(); err != nil {
		errs = append(errs, common.NewValidationError("password", err))
	}

	return errs
}

type AgencyLoginResponse struct {
	TFAToken AgencyTFAToken `json:"tfa_token"`
}

type AgencyTFARequest struct {
	TFAToken   AgencyTFAToken `json:"tfa_token"`
	TFACode    common.TFACode `json:"tfa_code"`
	RememberMe bool           `json:"remember_me"`
}

func (r AgencyTFARequest) Validate() []common.ValidationError {
	var errs []common.ValidationError

	if r.TFAToken == "" {
		errs = append(errs, common.NewValidationError("tfa_token", common.ErrRequired))
	}

	if r.TFACode == "" {
		errs = append(errs, common.NewValidationError("tfa_code", common.ErrRequired))
	} else if err := r.TFACode.Validate(); err != nil {
		errs = append(errs, common.NewValidationError("tfa_code", err))
	}

	return errs
}

type AgencyTFAResponse struct {
	SessionToken      AgencySessionToken  `json:"session_token"`
	PreferredLanguage common.LanguageCode `json:"preferred_language"`
	AgencyName        string              `json:"agency_name"`
}

// AgencyLogoutRequest is empty - session token passed via Authorization header
type AgencyLogoutRequest struct{}

// ============================================================================
// Agency User Invitation
// ============================================================================

var (
	errRolesRequired   = errors.New("at least one role is required")
	errRoleWrongPortal = errors.New("role does not belong to the agency portal")
)

type AgencyInviteUserRequest struct {
	EmailAddress        common.EmailAddress `json:"email_address"`
	InviteEmailLanguage common.LanguageCode `json:"invite_email_language,omitempty"`
	Roles               []common.RoleName   `json:"roles"`
}

func (r AgencyInviteUserRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError

	if err := r.EmailAddress.Validate(); err != nil {
		errs = append(errs, common.NewValidationError("email_address", err))
	}
	if r.InviteEmailLanguage != "" {
		if err := r.InviteEmailLanguage.Validate(); err != nil {
			errs = append(errs, common.NewValidationError("invite_email_language", err))
		}
	}

	if len(r.Roles) == 0 {
		errs = append(errs, common.NewValidationError("roles", errRolesRequired))
	} else {
		for _, role := range r.Roles {
			if err := role.Validate(); err != nil {
				errs = append(errs, common.NewValidationError("roles", common.ErrRoleNameInvalid))
			} else if !strings.HasPrefix(string(role), "agency:") {
				errs = append(errs, common.NewValidationError("roles", errRoleWrongPortal))
			}
		}
	}

	return errs
}

type AgencyInviteUserResponse struct {
	InvitationID string `json:"invitation_id"`
	ExpiresAt    string `json:"expires_at"`
}

type AgencyCompleteSetupRequest struct {
	InvitationToken   AgencyInvitationToken `json:"invitation_token"`
	Password          common.Password       `json:"password"`
	FullName          common.FullName       `json:"full_name"`
	PreferredLanguage common.LanguageCode   `json:"preferred_language,omitempty"`
}

func (r AgencyCompleteSetupRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError

	if r.InvitationToken == "" {
		errs = append(errs, common.NewValidationError("invitation_token", common.ErrRequired))
	}
	if err := r.Password.Validate(); err != nil {
		errs = append(errs, common.NewValidationError("password", err))
	}
	if err := r.FullName.Validate(); err != nil {
		errs = append(errs, common.NewValidationError("full_name", err))
	}
	if r.PreferredLanguage != "" {
		if err := r.PreferredLanguage.Validate(); err != nil {
			errs = append(errs, common.NewValidationError("preferred_language", err))
		}
	}

	return errs
}

type AgencyCompleteSetupResponse struct {
	Message string `json:"message"`
}

// ============================================
// User Management (Disable/Enable)
// ============================================

type AgencyDisableUserRequest struct {
	EmailAddress common.EmailAddress `json:"email_address"`
}

func (r AgencyDisableUserRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError

	if err := r.EmailAddress.Validate(); err != nil {
		errs = append(errs, common.NewValidationError("email_address", err))
	}

	return errs
}

type AgencyEnableUserRequest struct {
	EmailAddress common.EmailAddress `json:"email_address"`
}

func (r AgencyEnableUserRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError

	if err := r.EmailAddress.Validate(); err != nil {
		errs = append(errs, common.NewValidationError("email_address", err))
	}

	return errs
}

// ============================================
// Password Management
// ============================================

type AgencyPasswordResetToken = string

type AgencyRequestPasswordResetRequest struct {
	EmailAddress common.EmailAddress `json:"email_address"`
	Domain       common.DomainName   `json:"domain"`
}

func (r AgencyRequestPasswordResetRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError

	if r.EmailAddress == "" {
		errs = append(errs, common.NewValidationError("email_address", common.ErrRequired))
	} else if err := r.EmailAddress.Validate(); err != nil {
		errs = append(errs, common.NewValidationError("email_address", err))
	}

	if r.Domain == "" {
		errs = append(errs, common.NewValidationError("domain", common.ErrRequired))
	} else if err := r.Domain.Validate(); err != nil {
		errs = append(errs, common.NewValidationError("domain", err))
	}

	return errs
}

type AgencyRequestPasswordResetResponse struct {
	Message string `json:"message"`
}

type AgencyCompletePasswordResetRequest struct {
	ResetToken  AgencyPasswordResetToken `json:"reset_token"`
	NewPassword common.Password          `json:"new_password"`
}

func (r AgencyCompletePasswordResetRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError

	if r.ResetToken == "" {
		errs = append(errs, common.NewValidationError("reset_token", common.ErrRequired))
	}

	if r.NewPassword == "" {
		errs = append(errs, common.NewValidationError("new_password", common.ErrRequired))
	} else if err := r.NewPassword.Validate(); err != nil {
		errs = append(errs, common.NewValidationError("new_password", err))
	}

	return errs
}

type AgencyChangePasswordRequest struct {
	CurrentPassword common.Password `json:"current_password"`
	NewPassword     common.Password `json:"new_password"`
}

func (r AgencyChangePasswordRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError

	if r.CurrentPassword == "" {
		errs = append(errs, common.NewValidationError("current_password", common.ErrRequired))
	} else if err := r.CurrentPassword.Validate(); err != nil {
		errs = append(errs, common.NewValidationError("current_password", err))
	}

	if r.NewPassword == "" {
		errs = append(errs, common.NewValidationError("new_password", common.ErrRequired))
	} else if err := r.NewPassword.Validate(); err != nil {
		errs = append(errs, common.NewValidationError("new_password", err))
	}

	if r.CurrentPassword != "" && r.NewPassword != "" && r.CurrentPassword == r.NewPassword {
		errs = append(errs, common.NewValidationError("new_password", common.ErrNewPasswordSameAsCurrent))
	}

	return errs
}

// ============================================
// User Management (Filter Users)
// ============================================

type AgencyRole string

const (
	AgencyRoleViewUsers     AgencyRole = "view_users"
	AgencyRoleManageUsers   AgencyRole = "manage_users"
	AgencyRoleViewDomains   AgencyRole = "view_domains"
	AgencyRoleManageDomains AgencyRole = "manage_domains"
	AgencyRoleSuperadmin    AgencyRole = "superadmin"
)

type AgencyUser struct {
	EmailAddress common.EmailAddress `json:"email_address"`
	Name         string              `json:"name"`
	Status       string              `json:"status"`
	CreatedAt    string              `json:"created_at"`
	Roles        []AgencyRole        `json:"roles"`
}

type FilterAgencyUsersRequest struct {
	Limit        *int32  `json:"limit,omitempty"`
	Cursor       *string `json:"cursor,omitempty"`
	FilterEmail  *string `json:"filter_email,omitempty"`
	FilterName   *string `json:"filter_name,omitempty"`
	FilterStatus *string `json:"filter_status,omitempty"`
}

func (r FilterAgencyUsersRequest) Validate() []common.ValidationError {
	// Optional fields
	return nil
}

type FilterAgencyUsersResponse struct {
	Items      []AgencyUser `json:"items"`
	NextCursor string       `json:"next_cursor"`
}

// ===================================
// Language Management
// ===================================

type AgencySetLanguageRequest struct {
	Language common.LanguageCode `json:"language"`
}

func (r AgencySetLanguageRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError

	if r.Language == "" {
		errs = append(errs, common.NewValidationError("language", common.ErrRequired))
	} else if err := r.Language.Validate(); err != nil {
		errs = append(errs, common.NewValidationError("language", err))
	}

	return errs
}

// ===================================
// Get Current User Info
// ===================================

type AgencyMyInfoResponse struct {
	AgencyUserID      string              `json:"agency_user_id"`
	FullName          string              `json:"full_name"`
	PreferredLanguage common.LanguageCode `json:"preferred_language"`
	AgencyName        string              `json:"agency_name"`
	Roles             []string            `json:"roles"`
}

File: specs/typespec/agency/agency-users.ts
import {
	type EmailAddress,
	type Password,
	type FullName,
	type DomainName,
	type TFACode,
	type LanguageCode,
	type ValidationError,
	newValidationError,
	validateEmailAddress,
	validateEmployerEmail,
	validatePassword,
	validateFullName,
	validateDomainName,
	validateTFACode,
	validateLanguageCode,
	ERR_REQUIRED,
} from "../common/common";
import {
	type RoleName,
	type AssignRoleRequest,
	type RemoveRoleRequest,
	validateAssignRoleRequest,
	validateRemoveRoleRequest,
	validateRoleName,
	ERR_ROLE_NAME_INVALID,
} from "../common/roles";

export { validateAssignRoleRequest, validateRemoveRoleRequest };

// Token types
export type AgencySessionToken = string;
export type AgencyTFAToken = string;
export type DNSVerificationToken = string;
export type AgencySignupToken = string;
export type AgencyInvitationToken = string;
export type AgencyPasswordResetToken = string;

// ... (omitted sections)
// ============================================
// Signup Flow (DNS-based Domain Verification)
// ============================================

export interface AgencyInitSignupRequest {
	email: EmailAddress;
	home_region: string;
}

export function validateAgencyInitSignupRequest(
	request: AgencyInitSignupRequest
): ValidationError[] {
	const errs: ValidationError[] = [];

	if (!request.email) {
		errs.push(newValidationError("email", ERR_REQUIRED));
	} else {
		const emailErr = validateEmployerEmail(request.email);
		if (emailErr) {
			errs.push(newValidationError("email", emailErr));
		}
	}

	if (!request.home_region) {
		errs.push(newValidationError("home_region", ERR_REQUIRED));
	}

	return errs;
}

export interface AgencyGetSignupDetailsRequest {
	signup_token: string;
}

export function validateAgencyGetSignupDetailsRequest(
	request: AgencyGetSignupDetailsRequest
): ValidationError[] {
	const errs: ValidationError[] = [];

	if (!request.signup_token) {
		errs.push(newValidationError("signup_token", ERR_REQUIRED));
	}

	return errs;
}

/**
 * Request to complete agency signup after DNS verification.
 * The first user is automatically granted admin rights and assigned
 * both 'agency:invite_users' and 'agency:manage_users' roles.
 * All operations are atomic - either the entire signup succeeds or no data is created.
 */
export interface AgencyCompleteSignupRequest {
	signup_token: string;
	password: Password;
	preferred_language: LanguageCode;
	has_added_dns_record: boolean;
	agrees_to_eula: boolean;
}

export function validateAgencyCompleteSignupRequest(
	request: AgencyCompleteSignupRequest
): ValidationError[] {
	const errs: ValidationError[] = [];

	if (!request.signup_token) {
		errs.push(newValidationError("signup_token", ERR_REQUIRED));
	}

	if (!request.password) {
		errs.push(newValidationError("password", ERR_REQUIRED));
	} else {
		const passErr = validatePassword(request.password);
		if (passErr) {
			errs.push(newValidationError("password", passErr));
		}
	}

	if (!request.preferred_language) {
		errs.push(newValidationError("preferred_language", ERR_REQUIRED));
	} else {
		const langErr = validateLanguageCode(request.preferred_language);
		if (langErr) {
			errs.push(newValidationError("preferred_language", langErr));
		}
	}

	if (!request.has_added_dns_record) {
		errs.push(newValidationError("has_added_dns_record", ERR_REQUIRED));
	}

	if (!request.agrees_to_eula) {
		errs.push(newValidationError("agrees_to_eula", ERR_REQUIRED));
	}

	return errs;
}

// ============================================
// Login Flow
// ============================================

export interface AgencyLoginRequest {
	email: EmailAddress;
	domain: DomainName;
	password: Password;
}

export function validateAgencyLoginRequest(
	request: AgencyLoginRequest
): ValidationError[] {
	const errs: ValidationError[] = [];

	if (!request.email) {
		errs.push(newValidationError("email", ERR_REQUIRED));
	} else {
		const emailErr = validateEmailAddress(request.email);
		if (emailErr) {
			errs.push(newValidationError("email", emailErr));
		}
	}

	if (!request.domain) {
		errs.push(newValidationError("domain", ERR_REQUIRED));
	} else {
		const domainErr = validateDomainName(request.domain);
		if (domainErr) {
			errs.push(newValidationError("domain", domainErr));
		}
	}

	if (!request.password) {
		errs.push(newValidationError("password", ERR_REQUIRED));
	}

	return errs;
}

export interface AgencyTFARequest {
	tfa_token: string;
	tfa_code: TFACode;
	remember_me: boolean;
}

export function validateAgencyTFARequest(
	request: AgencyTFARequest
): ValidationError[] {
	const errs: ValidationError[] = [];

	if (!request.tfa_token) {
		errs.push(newValidationError("tfa_token", ERR_REQUIRED));
	}

	if (!request.tfa_code) {
		errs.push(newValidationError("tfa_code", ERR_REQUIRED));
	} else {
		const tfaErr = validateTFACode(request.tfa_code);
		if (tfaErr) {
			errs.push(newValidationError("tfa_code", tfaErr));
		}
	}

	return errs;
}
const ERR_ROLES_REQUIRED = "at least one role is required";
const ERR_ROLE_WRONG_PORTAL = "role does not belong to the agency portal";

export interface AgencyInviteUserRequest {
	email_address: EmailAddress;
	invite_email_language?: LanguageCode;
	roles: RoleName[];
}

export function validateAgencyInviteUserRequest(
	request: AgencyInviteUserRequest
): ValidationError[] {
	const errs: ValidationError[] = [];

	if (!request.email_address) {
		errs.push(newValidationError("email_address", ERR_REQUIRED));
	} else {
		const emailErr = validateEmailAddress(request.email_address);
		if (emailErr) {
			errs.push(newValidationError("email_address", emailErr));
		}
	}

	if (request.invite_email_language) {
		const langErr = validateLanguageCode(request.invite_email_language);
		if (langErr) {
			errs.push(newValidationError("invite_email_language", langErr));
		}
	}

	if (!request.roles || request.roles.length === 0) {
		errs.push(newValidationError("roles", ERR_ROLES_REQUIRED));
	} else {
		for (const role of request.roles) {
			const roleErr = validateRoleName(role);
			if (roleErr) {
				errs.push(newValidationError("roles", ERR_ROLE_NAME_INVALID));
			} else if (!role.startsWith("agency:")) {
				errs.push(newValidationError("roles", ERR_ROLE_WRONG_PORTAL));
			}
		}
	}

	return errs;
}

export interface AgencyInviteUserResponse {
	invitation_id: string;
	expires_at: string;
}

// ===================================
// Complete Setup
// ===================================

export interface AgencyCompleteSetupRequest {
	invitation_token: AgencyInvitationToken;
	password: Password;
	full_name: FullName;
	preferred_language?: LanguageCode;
}

export function validateAgencyCompleteSetupRequest(
	request: AgencyCompleteSetupRequest
): ValidationError[] {
	const errs: ValidationError[] = [];

	if (!request.invitation_token) {
		errs.push(newValidationError("invitation_token", ERR_REQUIRED));
	}

	if (!request.password) {
		errs.push(newValidationError("password", ERR_REQUIRED));
	} else {
		const passwordErr = validatePassword(request.password);
		if (passwordErr) {
			errs.push(newValidationError("password", passwordErr));
		}
	}

	if (!request.full_name) {
		errs.push(newValidationError("full_name", ERR_REQUIRED));
	} else {
		const fullNameErr = validateFullName(request.full_name);
		if (fullNameErr) {
			errs.push(newValidationError("full_name", fullNameErr));
		}
	}

	if (request.preferred_language) {
		const langErr = validateLanguageCode(request.preferred_language);
		if (langErr) {
			errs.push(newValidationError("preferred_language", langErr));
		}
	}

	return errs;
}

export interface AgencyCompleteSetupResponse {
	message: string;
}

// ... (omitted sections)

export interface AgencyUser {
	email_address: EmailAddress;
	name: string;
	status: string;
	created_at: string;
	roles: RoleName[];
}

export interface FilterAgencyUsersRequest {
	limit?: number;
	cursor?: string;
	filter_email?: string;
	filter_name?: string;
	filter_status?: string;
}

export function validateFilterAgencyUsersRequest(
	request: FilterAgencyUsersRequest
): ValidationError[] {
	const errs: ValidationError[] = [];
	// Optional fields, default validation
	return errs;
}

export interface FilterAgencyUsersResponse {
	items: AgencyUser[];
	next_cursor: string;
}

// ===================================
// Disable/Enable User
// ===================================

export interface AgencyDisableUserRequest {
	email_address: EmailAddress;
}

export function validateAgencyDisableUserRequest(
	request: AgencyDisableUserRequest
): ValidationError[] {
	const errs: ValidationError[] = [];

	if (!request.email_address) {
		errs.push(newValidationError("email_address", ERR_REQUIRED));
	} else {
		const emailErr = validateEmailAddress(request.email_address);
		if (emailErr) {
			errs.push(newValidationError("email_address", emailErr));
		}
	}

	return errs;
}

export interface AgencyEnableUserRequest {
	email_address: EmailAddress;
}

export function validateAgencyEnableUserRequest(
	request: AgencyEnableUserRequest
): ValidationError[] {
	const errs: ValidationError[] = [];

	if (!request.email_address) {
		errs.push(newValidationError("email_address", ERR_REQUIRED));
	} else {
		const emailErr = validateEmailAddress(request.email_address);
		if (emailErr) {
			errs.push(newValidationError("email_address", emailErr));
		}
	}

	return errs;
}

// ===================================
// Language Management
// ===================================

export interface AgencySetLanguageRequest {
	language: LanguageCode;
}

export function validateAgencySetLanguageRequest(
	request: AgencySetLanguageRequest
): ValidationError[] {
	const errs: ValidationError[] = [];

	if (!request.language) {
		errs.push(newValidationError("language", ERR_REQUIRED));
	} else {
		const langErr = validateLanguageCode(request.language);
		if (langErr) {
			errs.push(newValidationError("language", langErr));
		}
	}

	return errs;
}

// ===================================
// Get Current User Info
// ===================================

export interface AgencyMyInfoResponse {
	agency_user_id: string;
	full_name: string;
	preferred_language: LanguageCode;
	agency_name: string;
	roles: string[];
}
